### ISSUE TITLE
ExplicitOuter doesn't match Scala 2.12's outer types

### ISSUE THREAD

**Issue by gsps** [lampepfl/dotty#4342]
Scala 2.12 appears to use the declared self type of the enclosing class as the outer type when transforming inner classes. Dotty doesn't match this behavior and uses the enclosing class itself, leading to `NoSuchMethodErrors` wrt. Scala 2.12 classes upon executing the Dotty-compiled code. Here's a minimal example:

Consider the following code, compiled with 2.12:
```scala
package foo

class A extends B

trait B { this: A =>
  class C
}

object a extends A
```
and the following code, compiled with Dotty:
```scala
object Scala2_12FromDotty {
  new foo.a.C  // runtime error: trying to call constructor of C with a B instead of an A
}
```
We can see the discrepancy in the generated code. That is, `javap -p 'target/scala-2.12/classes/foo/B$C.class'` yields:
```
Compiled from "Foo.scala"
public class foo.B$C {
  public final foo.A $outer;
  public foo.A foo$B$C$$$outer();
  public foo.B$C(foo.A);
}

```
whereas `javap -c Scala2_12FromDotty$.class` yields
```
Compiled from "Scala2_12FromDotty.scala"
public final class Scala2_12FromDotty$ {
  (...)
  public Scala2_12FromDotty$();
    Code:
      (...)
      12: getstatic     #25                 // Field foo/a$.MODULE$:Lfoo/a$;
      15: invokespecial #28                 // Method foo/B$C."<init>":(Lfoo/B;)V
      (...)
}
```

I tried to look into the related phases a bit, but I'm not sure whether this should be fixed in `ExplicitOuter` or by transforming infos, applications and defdefs later, for instance, as an extension of `LinkScala2Impls`.

**Comment by smarter** [comment]
I don't have much knowledge of the relevant code here, but this looks similar to what https://github.com/lampepfl/dotty/blob/2212dc0ea072614fabda1004d9cbb23401cba351/compiler/src/dotty/tools/dotc/transform/LinkScala2Impls.scala#L51-L52 tries to address.

**Comment by gsps** [comment]
@smarter Yeah, I was looking at that part. I'll just take a stab at it, then.
